geode-core/src/test/java/org/apache/geode/internal/cache/CacheWriterGetOldValueIntegrationTest.java

```diff
@@ -0,0 +1,286 @@
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.internal.cache;
+
+import static org.assertj.core.api.Assertions.*;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+
+import org.apache.logging.log4j.Logger;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import org.apache.geode.cache.CacheFactory;
+import org.apache.geode.cache.CacheWriterException;
+import org.apache.geode.cache.DataPolicy;
+import org.apache.geode.cache.EntryEvent;
+import org.apache.geode.cache.EvictionAction;
+import org.apache.geode.cache.EvictionAttributes;
+import org.apache.geode.cache.Region;
+import org.apache.geode.cache.RegionFactory;
+import org.apache.geode.cache.util.CacheWriterAdapter;
+import org.apache.geode.internal.cache.GemFireCacheImpl;
+import org.apache.geode.internal.logging.LogService;
+import org.apache.geode.test.junit.categories.IntegrationTest;
+
+@Category(IntegrationTest.class)
+public class CacheWriterGetOldValueIntegrationTest {
+
+  private GemFireCacheImpl cache = null;
+
+  @Before
+  public void setUp() throws Exception {
+    createCache();
+  }
+
+  private void createCache() {
+    cache =
+        (GemFireCacheImpl) new CacheFactory().set("locators", "").set("mcast-port", "0").create();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    cache.close();
+    expectedValues.clear();
+  }
+
+  @Test
+  public void getOldValueInCacheWriterReturnsValueOfEvictedEntry() {
+    doTest(false);
+  }
+
+  @Test
+  public void getOldValueWithTransactionInCacheWriterReturnsValueOfEvictedEntry() {
+    doTest(true);
+  }
+
+  @Test
+  public void doPutAll() {
+    PutAllCacheWriter<String, String> cw;
+    Region<String, String> r = createOverflowRegion();
+    put(r, "k1", "v1");
+    put(r, "k2", "v2");
+    cw = new PutAllCacheWriter<>();
+    r.getAttributesMutator().setCacheWriter(cw);
+    HashMap<String, String> putAllMap = new HashMap<>();
+    putAllMap.put("k1", "update1");
+    putAllMap.put("k2", "update2");
+    r.putAll(putAllMap);
+
+    assertThat(cw.getSeenEntries()).isEqualTo(this.expectedValues);
+  }
+
+  @Test
+  public void doRemoveAll() {
+    RemoveAllCacheWriter<String, String> cw;
+    Region<String, String> r = createOverflowRegion();
+    put(r, "k1", "v1");
```

Comment:
Can we change the name to populate region and set expected values (other than put) ...I am reading this as region.put() not realizing its a test method...